plugins/pixart-rf/fu-pxi-device.c

```diff
@@ -0,0 +1,611 @@
+
+	/* read fw ota init new command */
+	res[0] = PXI_HID_DEV_OTA_FEATURE_REPORT_ID;
+	res[1] = FU_PXI_DEVICE_CMD_FW_OTA_INIT_NEW;
+	if (!fu_pxi_device_get_feature (self, res, sizeof(res), error))
+		return FALSE;
+
+	/* shared state */
+	if (!fu_common_read_uint8_safe (res, sizeof(res), 0x3,
+					&self->status, error))
+		return FALSE;
+	if (!fu_common_read_uint8_safe (res, sizeof(res), 0x4,
+					&self->new_flow, error))
+		return FALSE;
+	if (!fu_common_read_uint16_safe (res, sizeof(res), 0x5,
+					 &self->offset, G_LITTLE_ENDIAN, error))
+		return FALSE;
+	if (!fu_common_read_uint16_safe (res, sizeof(res), 0x7,
+					 &self->checksum, G_LITTLE_ENDIAN, error))
+		return FALSE;
+	if (!fu_common_read_uint32_safe (res, sizeof(res), 0x9,
+					 &self->max_object_size, G_LITTLE_ENDIAN, error))
+		return FALSE;
+	if (!fu_common_read_uint16_safe (res, sizeof(res), 0xd,
+					 &self->mtu_size, G_LITTLE_ENDIAN, error))
+		return FALSE;
+	if (!fu_common_read_uint16_safe (res, sizeof(res), 0xf,
+					 &self->prn_threshold, G_LITTLE_ENDIAN, error))
+		return FALSE;
+	if (!fu_common_read_uint8_safe (res, sizeof(res), 0x11,
+					&self->spec_check_result, error))
+		return FALSE;
+
+	/* sanity check */
+	if (self->spec_check_result != OTA_SPEC_CHECK_OK) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_READ,
+			     "FwInitNew Spec Check Fail 0x%02x",
+			     self->spec_check_result);
+		return FALSE;
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_pxi_device_fw_upgrade (FuPxiDevice *self, FuFirmware *firmware, GError **error)
+{
+	const gchar *version;
+	const guint8 *buf;
+	gsize bufsz = 0;
+	guint8 fw_version[10] = { 0x0 };
+	guint8 opcode = 0;
+	guint16 checksum;
+	g_autoptr(GBytes) fw = NULL;
+	g_autoptr(GByteArray) req = g_byte_array_new ();
+
+	fw = fu_firmware_get_image_default_bytes (firmware, error);
+	if (fw == NULL)
+		return FALSE;
+	buf = g_bytes_get_data (fw, &bufsz);
+	checksum = fu_pxi_device_calculate_checksum (buf, bufsz);
+	fu_byte_array_append_uint8 (req, PXI_HID_DEV_OTA_OUTPUT_REPORT_ID);
+	fu_byte_array_append_uint8 (req, FU_PXI_DEVICE_CMD_FW_UPGRADE);
+	fu_byte_array_append_uint32 (req, bufsz, G_LITTLE_ENDIAN);
+	fu_byte_array_append_uint16 (req, checksum, G_LITTLE_ENDIAN);
+	version = fu_firmware_get_version (firmware);
+	if (!fu_memcpy_safe (fw_version, sizeof(fw_version), 0x0,	/* dst */
+			     (guint8 *) version, strlen (version), 0x0,	/* src */
+			     strlen (version), error))
+		return FALSE;
+	g_byte_array_append (req, fw_version, sizeof(fw_version));
+
+	/* send fw upgrade command */
+	fu_device_set_status (FU_DEVICE (self), FWUPD_STATUS_DEVICE_VERIFY);
+	if (!fu_udev_device_pwrite_full (FU_UDEV_DEVICE (self), 0,
+					 req->data, req->len, error))
+		return FALSE;
+	if (g_getenv ("FWUPD_PIXART_RF_VERBOSE") != NULL)
+		fu_common_dump_raw (G_LOG_DOMAIN, "fw upgrade", req->data, req->len);
+
+	/* read fw upgrade command result */
+	g_debug ("read fw upgrade result");
```

Comment:
this seems like an unnecessary debug line

Reply:
Removed.